### DISCUSSION TITLE
Monorepo structure for frontend and backend packages and libraries

### DISCUSSION THREAD

**Discussion by christiankaseburg** [nrwl/nx#12065] [General]
Hey all,

I have a few questions about a monorepo structure with NX. I have started a small project that I have two web apps, a few APIs, and common shared libraries the apps and APIs use. I'm coming across a scenario that I feel that I am coupling the frontend and backend together and it may be a good time to separate frontend from the backend as well as the shared libraries as my personal project grows. In my head I believe it makes since to create a libs folder for shared backend code, a shared folder for frontend code, and separate app directories for frontend and backend services.

Note: All projects are either Typescript or Go. I do not believe that it matters here, but figured I would point it out.

Here is an example of how my mono repository is currently:
```
packages/
  webApp1/
  webApp2,
  apiA/
  apiB/
  apiC/
libs/
  core/ - All APIs have a dependency on this project, but none of the webApp's do.
  webAssets/  - All web apps have a dependency on this but APIs do not.
  sharedLibA/
  sharedLibB/
  ```
  
I'm thinking of moving to more of a structure like this:
```
  frontendPackages/
  webApp1/
  webApp2/
backendPackages/
  apiA/
  apiB/
  apiC/
frontendLibs/
  webAssets/ 
backendLibs/
  core/
commonlibs/ - This could be common between frontend and backend
  sharedLibA/
  sharedLibB/
```
  
Let me know your thoughts or feedback on seperating frontend and backend in a monorepo.

**Comment by mend3** [comment]
Actually, it doesn't matter that much where the files are. The point is, in my opinion, you should have a pattern for your namings that is easy for you to refer in code and follow your structure logic.

I believe at some points, too much nested folders are worse, once the code only grows forward.

This is my structure right now.

- apps
- - api (back)
- - app (front)
- - storybook (front)
- packages
- - db (database types)
- - logger
- - shared
- - - instances (both, wrappers, singletons, etc)
- - - ui (front, publishable isolated components)
- - utils

![image](https://github.com/nrwl/nx/assets/1450229/2f9e211e-2b54-41ca-ad7c-84c40fdd25ea)
